### PR TITLE
DuckDNS.org and Name.com DNS API support

### PR DESCRIPTION
**Pull request by raidenii**
DuckDNS.org and Name.com DNS API support. Tested working.

### CODE REVIEW

**Comment by Neilpang** [comment]
DuckDNS.org  只能 使用 `*.DuckDNS.org` 的二级域名吗?  还是可以支持自定义的顶级域名 ?

File: dnsapi/dns_duckdns.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+########  Public functions #####################
+
+#Usage: dns_duckdns_add _acme-challenge.domain.duckdns.org "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_duckdns_add() {
+  fulldomain=$1
+  txtvalue=$2
+
+  # We'll extract the domain/username from full domain
+  DuckDNS_domain=$(printf "%s\n" "$fulldomain" | rev | cut -d . -f 3 | rev)
```
Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: 不要使用 `rev`

File: dnsapi/dns_duckdns.sh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  if [ -z "$DuckDNS_domain" ]; then
+    _err "Error extracting the domain."
+    return 1
+  fi
+
+  if [ -z "$DuckDNS_token" ]; then
```
Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: 这个名字为什么不用 大写:
```
DuckDNS_Token
```

File: dnsapi/dns_namecom.sh [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+namecom_api="https://api.name.com/api"
+
+#Usage: dns_namecom_add   _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_namecom_add() {
+  fulldomain=$1
+  txtvalue=$2
+
+  # First we need name.com credentials.
+  if [ -z "$namecom_username" ]; then
```
Comment (Neilpang) [inline dnsapi/dns_namecom.sh]: 这个名字需要大写

File: dnsapi/dns_namecom.sh [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    _err "Please specify that in your environment variable."
+    return 1
+  fi
+
+  # Save them in configuration.
+  _saveaccountconf namecom_username "$namecom_username"
+  _saveaccountconf namecom_token "$namecom_token"
+
+  # Login in using API
+  _namecom_login
```
Comment (Neilpang) [inline dnsapi/dns_namecom.sh]: 这个方法不需要检测返回值吗

File: dnsapi/dns_namecom.sh [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    _err "Unable to find domain specified."
+    _namecom_logout
+    return 1
+  fi
+
+  # Add TXT record.
+  _namecom_addtxt_json="{\"hostname\":\"$_sub_domain\",\"type\":\"TXT\",\"content\":\"$txtvalue\",\"ttl\":\"300\",\"priority\":\"10\"}"
+  if _namecom_rest POST "dns/create/$_domain" "$_namecom_addtxt_json"; then
+    retcode=$(printf "%s\n" "$response" | _egrep_o "\"code\":100")
+    if [ ! -z "$retcode" ]; then
```
Comment (Neilpang) [inline dnsapi/dns_namecom.sh]: 直接这样写
```
if [ "$retcode" ]; then
```

File: dnsapi/dns_namecom.sh [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  else
+    response="$(_get "$namecom_api/$param")"
+  fi
+
+  if [ "$?" != "0" ]; then
+    _err "error $param"
+    return 1
+  fi
+
+  _debug response "$response"
```
Comment (Neilpang) [inline dnsapi/dns_namecom.sh]: _debug2 response

File: dnsapi/dns_duckdns.sh (hunk 1 above)
Comment (raidenii) [inline dnsapi/dns_duckdns.sh]: 不用rev还真不好提取子域名……

File: dnsapi/dns_namecom.sh (hunk 4 above)
Comment (raidenii) [inline dnsapi/dns_namecom.sh]: login返回1的话登录失败

**Comment by raidenii** [comment]
Duckdns是提供动态域名的，只能用二级域名，见spec：
https://www.duckdns.org/spec.jsp

File: dnsapi/dns_duckdns.sh (hunk 1 above)
Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]: 用 sed

Comment (raidenii) [inline dnsapi/dns_duckdns.sh]: 之前用过sed，不过要用n参数，表达式里面加上p来控制输出，acme.sh貌似不支持这么做

Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]:
```sh
DuckDNS_domain=$(echo "$fulldomain" | _lower_case | sed 's/.duckdns.org//' )
```

Comment (raidenii) [inline dnsapi/dns_duckdns.sh]: 这样输出的结果不对。fulldomain一般是_acme-challenge.xxx.duckdns.org，包含子域名的情况下是_acme-challenge.yyy.xxx.duckdns.org，提取domain的时候应该提取的是xxx，即使再加上sed 's/_acme-challenge//' 也不一定能够提取出正确的域名。

Comment (Neilpang) [inline dnsapi/dns_duckdns.sh]:
```sh
DuckDNS_domain=$(echo $fulldomain | _lower_case | _egrep_o '.[^.]*.duckdns.org' | cut -d . -f 2)
```

File: dnsapi/dns_namecom.sh [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    _err "Please specify that in your environment variable."
+    return 1
+  fi
+
+  # Save them in configuration.
+  _saveaccountconf Namecom_Username "$Namecom_Username"
+  _saveaccountconf Namecom_Token "$Namecom_Token"
+
+  # Login in using API
+  _namecom_login
```
Comment (Neilpang) [inline dnsapi/dns_namecom.sh]: 这里为什么不加上一句 判断返回值的

**Comment by Neilpang** [comment]
请更新这两个 readme 文件: 
https://github.com/Neilpang/acme.sh/blob/master/README.md
https://github.com/Neilpang/acme.sh/blob/master/dnsapi/README.md

添加到支持列表, 并写上用法, 步骤.

**Comment by RockyTV** [comment]
May I ask why was this user's DuckDNS support chosen over mine?

**Comment by Neilpang** [comment]
Sorry,  @RockyTV it seems that you didn't respond to my review for a long time, I just forgot about your PR.

And when I saw this guy's PR, I didn't remember you had one. 

There is no reason to choose or not,  I just saw this PR, and reviewed the code. And it's all ready to merge. 

I just checked your PR again just now,  I didn't even notice that this code was forked from yours.

But sorry,  I didn't meant to not choose your code.

Thanks anyway.

**Comment by raidenii** [comment]
I have to say that, my code was not forked from RockyTV's. I wrote it on my own. You can tell from the coding style and ideas behind it - they are totally different. I did not check the PRs until I have forked acme.sh, uploaded my version and then saw yours.

**Pull request merged**